src/code/containers/challenge-container-selector.js

```diff
@@ -43,27 +44,33 @@ class ChallengeContainerSelector extends Component {
   }
 
   componentWillMount() {
-    const { routeParams, navigateToCurrentRoute, navigateToChallenge } = this.props;
-    if (hasChangedRouteParams(this.props)) {
+    const { navigateToCurrentRoute, navigateToChallenge, authoring } = this.props;
+    // the URL's challengeId is only used for initial routing, so prioritize the numeric route params
```

Comment:
I don't quite understand having the challengeId be in the urlParams and the routeParams be in the hash. It seems to me that as the user moves forward, there will be a mismatch between challengeId and the routeSpec, which will be confusing. After finishing `?challengeId=basicsIntroVisible` I will go to `?challengeId=basicsIntroVisible#/2/1/2`, which doesn't seem right.

I think what would make more sense for the user would be to be able to go to `#/basicsIntroVisible` (which is also easier to remember, because you don't need the `challengeId=`), have that load up, and then if you continue then your url changes to `#/2/1/2`.

It looks like this wouldn't be too hard. I tested the following, but I may have missed something:

Change app.js to include the new route definition:
```
<Router history={history}>
  <Route path="/(:level/:mission/:challenge)" component={ChallengeContainerSelector} />
  <Route path="/(:challengeId)" component={ChallengeContainerSelector} />
</Router>
```
Change challenge-container-selector.js to look for the challengeId in the routeParams:
```
componentWillMount() {
    const { navigateToCurrentRoute, navigateToChallenge, authoring } = this.props;
    let routeParams = this.props.routeParams;
    if (routeParams.challengeId) {
      routeParams = AuthoringUtils.challengeIdToRouteParams(authoring, routeParams.challengeId);
    }
    if (hasChangedRouteParams(this.props.currentRouteSpec, routeParams)) {
      navigateToCurrentRoute({level: routeParams.level-1, mission: routeParams.mission-1, challenge: routeParams.challenge-1});
    } else {
      navigateToChallenge({level: 0, mission: 0, challenge: 0});
    }
  }
```
Does that make sense?